src/main/java/im/status/keycard/KeycardApplet.java

```diff
@@ -882,15 +886,25 @@ private void loadData(APDU apdu) {
     if (!pin.isValidated()) {
       ISOException.throwIt(ISO7816.SW_CONDITIONS_NOT_SATISFIED);
     }
-    
-    if ((short) apduBuffer[ISO7816.OFFSET_LC] != SAVED_DATA_SIZE) {
+
+    // Included data must be a full page
+    if (apduBuffer[ISO7816.OFFSET_LC] != SAVED_DATA_PAGE_SZ) {
+      ISOException.throwIt(ISO7816.SW_WRONG_DATA);
+    }
+
+    // Must reference a page in the saved data boundary (0 - (SAVED_DATA_NUM_PAGES-1))
+    if ((short) apduBuffer[ISO7816.OFFSET_P1] >= SAVED_DATA_NUM_PAGES) {
       ISOException.throwIt(ISO7816.SW_WRONG_DATA);
     }
     
+    // Copy this page of data
     JCSystem.beginTransaction();
-    Util.arrayCopy(apduBuffer, (short) ISO7816.OFFSET_CDATA, savedData, (short) 0, apduBuffer[ISO7816.OFFSET_LC]);
+    Util.arrayCopy( apduBuffer, 
+                    (short) ISO7816.OFFSET_CDATA, 
+                    savedData, 
+                    (short) (apduBuffer[ISO7816.OFFSET_P1] * SAVED_DATA_PAGE_SZ), 
+                    apduBuffer[ISO7816.OFFSET_LC]);
```

Comment:
Should be switched to `SAVED_DATA_PAGE_SZ` for readability

Reply:
[6963688](https://github.com/GridPlus/safe-card/pull/57/commits/696368846406b344e25b21023b42c61a5907f36f)